packages/core/src/sync/index.ts

```diff
@@ -431,483 +308,487 @@ export const createSync = async (args: CreateSyncParameters): Promise<Sync> => {
-            const { events, cursor } = await args.syncStore.getEvents({
-              filters: args.indexingBuild.sources.map(({ filter }) => filter),
-              from,
-              to: to < estimatedTo ? to : estimatedTo,
-              limit: getEventsMaxBatchSize,
-            });
+        const localEventGenerator = getLocalEventGenerator({
+          common: params.common,
+          network,
+          syncStore: params.syncStore,
+          sources,
+          localSyncGenerator,
+          from:
+            params.initialCheckpoint !== ZERO_CHECKPOINT_STRING
+              ? params.initialCheckpoint
+              : getChainCheckpoint({ syncProgress, network, tag: "start" })!,
+          to,
+          limit: Math.round(
+            params.common.options.syncEventsQuerySize /
+              (params.indexingBuild.networks.length * 2),
+          ),
+        });
 
-            args.common.logger.debug({
-              service: "sync",
-              msg: `Fetched ${events.length} events from the database for a ${formatEta(estimateSeconds * 1000)} range from timestamp ${decodeCheckpoint(from).blockTimestamp}`,
-            });
+        return bufferAsyncGenerator(localEventGenerator, 1);
+      },
+    );
 
-            for (const network of args.indexingBuild.networks) {
-              updateHistoricalStatus({ events, checkpoint: cursor, network });
-            }
+    const mergeAsync =
+      params.mode === "multichain"
+        ? mergeAsyncGenerators
+        : mergeAsyncGeneratorsWithEventOrder;
+
+    for await (const { events, checkpoint } of mergeAsync(eventGenerators)) {
+      if (params.mode === "multichain") {
+        const network = params.indexingBuild.networks.find(
+          (network) =>
+            network.chainId === Number(decodeCheckpoint(checkpoint).chainId),
+        )!;
+        params.common.logger.debug({
+          service: "sync",
+          msg: `Sequenced ${events.length} '${network.name}' events for timestamps [${decodeCheckpoint(cursor).blockTimestamp}, ${decodeCheckpoint(checkpoint).blockTimestamp}]`,
+        });
+      } else {
+        params.common.logger.debug({
+          service: "sync",
+          msg: `Sequenced ${events.length} events for timestamps [${decodeCheckpoint(cursor).blockTimestamp}, ${decodeCheckpoint(checkpoint).blockTimestamp}]`,
+        });
+      }
 
-            estimateSeconds = estimate({
-              from: decodeCheckpoint(from).blockTimestamp,
-              to: decodeCheckpoint(cursor).blockTimestamp,
-              target: getEventsMaxBatchSize,
-              result: events.length,
-              min: 10,
-              max: 86_400,
-              prev: estimateSeconds,
-              maxIncrease: 1.08,
-            });
+      for (const network of params.indexingBuild.networks) {
+        updateHistoricalStatus({ events, checkpoint, network });
+      }
+      yield events;
+      cursor = checkpoint;
+    }
+  }
+
+  const getOnRealtimeSyncEvent = () => {
```

Comment:
Do you need this closure / scope or could the `checkpoints` and `latencyTimers` move to the upper scope and it could just be `const onRealtimeSyncEvent = (event: ...) => { ... }`